app/models/concerns/preserve_order.rb

```diff
@@ -0,0 +1,13 @@
+module PreserveOrder
```

Comment:
I get the whole concern thing and code reuse and all, but find the the indirection here just ends up confusing things, I would just duplicate this twice.

Also the way it takes in a Hash and only looks at the first element feels risky to me.